File: call_api.js
const axios = require('axios');

/**
 * Calls the OpenRouter API with a conversation and config
 * @param {Array} conversation - Array of message objects with role and content
 * @param {Object} config - Configuration object containing API settings
 * @returns {Promise<string>} - The assistant's response content
 */
async function call_api(conversation, config) {
  try {
    const apiKey = config.api_key || process.env.API_KEY;
    
    if (!apiKey) {
      throw new Error("API key is required. Set it in config.js or as API_KEY environment variable.");
    }

    // Convert the conversation to the OpenRouter format
    const messages = conversation.map(msg => ({
      role: msg.role,
      content: msg.content
    }));

    const response = await axios({
      method: 'post',
      url: config.endpoint,
      headers: {
        'Content-Type': 'application/json',
        'Authorization': `Bearer ${apiKey}`,
        'HTTP-Referer': 'https://localhost:3000', // Replace with your actual domain
        'X-Title': 'Agent Runner'
      },
      data: {
        model: config.default_model,
        messages: messages,
        max_tokens: 4000
      }
    });

    // Extract the assistant's response content from OpenRouter format
    if (response.data && response.data.choices && response.data.choices.length > 0) {
      return response.data.choices[0].message.content;
    } else {
      throw new Error("Unexpected API response format");
    }
  } catch (error) {
    console.error("API call failed:", error.response?.data || error.message);
    throw new Error(`API call failed: ${error.message}`);
  }
}

module.exports = { call_api };

File: AGENT_WITH_PYTHON_SCRIPT_IN_JSON/call_api.py
import subprocess
import json

def call_api(conversation, config):
    """
    Call the API with the given conversation and configuration.
    Parameters:
        conversation (list): The list of messages to send to the API.
        config (dict): API configuration (endpoint, api_key, default_model).
    Returns:
        str: The response content from the API.
    """
    payload = {
        "model": config["default_model"],
        "messages": conversation
    }

    # Safely encode the payload to JSON
    payload_str = json.dumps(payload).replace("'", "'\\''")

    curl_command = f"""curl {config["endpoint"]} \\
      -H "Authorization: Bearer {config["api_key"]}" \\
      -H "Content-Type: application/json" \\
      -d '{payload_str}'"""

    output_file = "temp_output.json"
    try:
        subprocess.run(curl_command + f" -o {output_file}", shell=True, check=True)
        
        with open(output_file, 'r', encoding='utf-8') as f:
            response = f.read()

        json_response = json.loads(response)

        content = json_response.get('content', '') or \
                  json_response.get('choices', [{}])[0].get('message', {}).get('content', '')

        return content

    except Exception as e:
        print(f"Error calling API: {e}")
        return f"Error: {str(e)}"
